### DISCUSSION TITLE
Trimming a non-uniform background in book scans, where `-fuzz` fails

### DISCUSSION THREAD

**Discussion by ghost** [ImageMagick/ImageMagick#6901] [Help] [answered]
### ImageMagick version

7.1.1-21

### Operating system, version and so on

macOS

### Description

Hello,

I have book scans (not done by me), and the background is somewhat complex, i.e., grayish black but with small white dots. See sample image:

![scanned_input](https://github.com/ImageMagick/ImageMagick/assets/146453846/e1475a53-90ad-41dd-a0ee-e378abf8ca40)

Normally, for uniform backgrounds, I add a black border using `-bordercolor black`, then I do a combination of `-fuzz` and `-trim`, something like:

`convert scanned_input.png -bordercolor black -border 20 -fuzz 80% -trim +repage scanned_output.png`

But in this case, the background is non-uniform, and `-fuzz` either trims too much or not enough. How would you approach this in ImageMagick?

Thank you in advance!

**Comment by snibgo** [comment]
I can see the image but for some reason I can't download it.

I would denoise the image before trimming. The denoising might be with `-morphology`, eroding small light areas. Other possibilities are `-connected-components`, or `-statistic Median`. In any case, I would process the image (simplifying it) to find the crop I want, and then apply that crop to the original image.

**Comment by fmw42** [comment]
Threshold the image. Then use morphology open and close to remove most of the noise and fill in the page area. Then use connected components to clean up further.  Then use -format %@ to get the trim box or get the bounding box from the connected components.  Then use one of those to crop the original image.

**Comment by ghost** [comment]
Thank you both! You were extremely helpful. I tried implementing your recommendations, and came up with the following commands so far:

```
% convert scannedinput.png -threshold 50% -morphology Open Disk -morphology Close Disk -fuzz 10% -format "%@" info:
985x1635+166+147
% convert scannedinput.png -crop 985x1635+166+147 scannedoutput.png
```

producing the following result:

![scanned_output](https://github.com/ImageMagick/ImageMagick/assets/146453846/8aff51b6-14f1-4139-b224-f1c22bfd21b5)

This is pretty damn good! Probably good enough... But tell me if I’ve made mistakes. Also, I couldn’t understand the `-connected-components` that you both mentioned, despite reading the documentation at:

https://imagemagick.org/script/connected-components.php

**Comment by snibgo** [comment]
When an image has been simplified to a small number of flat colours, `-connected-components` can merge small components into larger ones, and find the largest component, and a few other tricks. If we squint at your image, we can see it basically has one large component (the scanned page) surrounded by smaller components (other pages, and the black background). For such images, `-connected-components` can be used to find the bounding box of components.

But you have found a solution that doesn't need `-connected-components`.

**Comment by fmw42** [comment]
You can do better if you use a lower threshold and check your threshold result and also use a smaller morphology kernel and check that result (by adding +write tempX.png after each step).  Also on IM 7, I suggest using magick in place of convert.

```
cropbox=`magick scannedinput.png \
-threshold 25% \
-morphology open diamond:2 \
+write threshold.png -format %@ info:`
magick scannedinput.png \
-crop $cropbox +repage result.png

```

Threshold and Morphology Image:

![threshold](https://github.com/ImageMagick/ImageMagick/assets/13458218/56ec82a3-2e1f-427c-9f7c-4fdbf90bd361)

Result Image:

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/afa27b7f-1eac-4881-8dce-d8eb3d82c51f)

**Reply by fmw42** [reply]
Or Alternately,

```
cropbox=`magick scannedinput.png \
-threshold 25% \
-morphology open diamond:2 \
-type bilevel \
+write threshold.png \
-define connected-components:area-threshold=10000 \
-define connected-components:mean-color=true \
-define connected-components:exclude-header=true \
-define connected-components:keep-top=2 \
-define connected-components:verbose=true \
-connected-components 8 cleaned.png | grep "gray(255)" | awk '{print $2}'`
magick scannedinput.png \
-crop $cropbox +repage result.png

```

```
This defines an area-threshold to merge if smaller than threshold of 10,000 pixels (adjust as desired)
It set the mean-color to true (for the cleaned image rather than its id values)
It disables the headers in the output
It keeps only the largest two regions (one black and one white)
It outputs the bounding box information from verbose=true
It saves the cleaned.png image.
It passes to grep to select only the gray(255) region (white)
It passes to awk to output the bounding box only from other information.
The resulting bounding box is put into the cropbox variable.

Then the dropbox variable is used to crop the input.
```

Cleaned Image:

![cleaned](https://github.com/ImageMagick/ImageMagick/assets/13458218/7e9bf312-514a-47a3-87eb-05ce1baf1c4b)

Result Image:

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/db66372b-975d-4d1c-aa50-70cb124e2b7b)

**Reply by fmw42** [reply, marked as answer]
Or adding morphology close

```
cropbox=`magick scannedinput.png \
-threshold 25% \
-morphology open diamond:2 -morphology close square:9 \
-type bilevel \
+write threshold.png \
-define connected-components:area-threshold=10000 \
-define connected-components:mean-color=true \
-define connected-components:exclude-header=true \
-define connected-components:keep-top=2 \
-define connected-components:verbose=true \
-connected-components 8 cleaned.png | grep "gray(255)" | awk '{print $2}'`
magick scannedinput.png \
-crop $cropbox +repage result.png
```

Cleaned Image:

![cleaned](https://github.com/ImageMagick/ImageMagick/assets/13458218/2323688c-bc32-4e7d-938c-cdad84afa5ef)

Result Image:

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/cbc7b1f3-1e4e-4946-9f47-d9f441eb3645)

**Comment by ghost** [comment]
Thank you both again! Ahhh... `keep-top=2` is the key component I was missing in connecting the dots and understanding `connected-components`. It’s barely mentioned in the documentation. I will sleep less ignorant this weekend. Problem solved!
